robotbase/src/main/java/com/team2357/frc2022/subsystems/TurretSubsystem.java

```diff
@@ -0,0 +1,241 @@
+package com.team2357.frc2022.subsystems;
+
+import com.revrobotics.CANSparkMax;
+import com.revrobotics.SparkMaxPIDController;
+import com.revrobotics.CANSparkMax.IdleMode;
+import com.team2357.frc2022.Constants;
+import com.team2357.frc2022.util.VisionTargetSupplier;
+import com.team2357.lib.arduino.ArduinoUSBController;
+import com.team2357.lib.subsystems.ClosedLoopSubsystem;
+import com.team2357.lib.subsystems.LimelightSubsystem.VisionTarget;
+
+import edu.wpi.first.wpilibj.smartdashboard.SmartDashboard;
+
+public class TurretSubsystem extends ClosedLoopSubsystem {
+    CANSparkMax m_turretMotor;
+    private SparkMaxPIDController m_pidController;
+    private ArduinoUSBController m_arduinoHallEffectSensor;
+    Configuration m_config;
+
+    private VisionTargetSupplier m_targetSupplier;
+    private VisionTarget m_currentTarget;
+
+    private double m_setPoint;
+
+    // Has the turret been zeroed
+    private boolean m_isZeroed;
+
+    public static class Configuration {
+        public IdleMode turretMotorIdleMode = IdleMode.kBrake;
+        public int m_turretMotorStallLimitAmps = 0;
+        public int m_turretMotorFreeLimitAmps = 0;
+
+        public double m_turretMotorP = 0;
+        public double m_turretMotorI = 0;
+        public double m_turretMotorD = 0;
+        public double m_turretMotorIZone = 0;
+        public double m_turretMotorFF = 0;
+        public double m_turretMotorMaxOutput = 0;
+        public double m_turretMotorMinOutput = 0;
+        public double m_turretMotorMaxRPM = 0;
+        public double m_turretMotorMaxVel = 0;
+        public double m_turretMotorMinVel = 0;
+        public double m_turretMotorMaxAcc = 0;
+        public double m_turretMotorAllowedError = 0;
+
+        public double m_turretRotationsClockwiseSoftLimit = 0;
+        public double m_turretRotationsCounterClockwiseSoftLimit = 0;
+        public double m_rotationsPerDegree = 0;
+    }
+
+    public TurretSubsystem(CANSparkMax turretMotor) {
+        m_turretMotor = turretMotor;
+
+        m_arduinoHallEffectSensor = new ArduinoUSBController(Constants.ARDUINO.ARDUINO_SENSOR_DEVICE_NAME);
+
+        m_arduinoHallEffectSensor.start();
```

Comment:
We probably want the USB to be its own subsystem and control its own lifecycle, not here.

Reply:
I agree, that will be a good thing for the freshmen to do. I did not think much of it when I added it since we were not sure how many Arduinos we were going to use yet.